File: code/sorting/Sleep_Sort/Sleep_Sort.cs
/* Part of Cosmos by OpenGenus Foundation */
using System;
using System.Threading;
using System.Linq;

namespace dotnetcore
{
  class Program
  {
    static void Main(string[] args)
    {
      SleepSort(new int[] {5, 4, 1, 2, 3});
    }

    static void SleepSort(int[] data)
    {
      data.AsParallel().WithDegreeOfParallelism(data.Length).Select(SleepAndReturn).ForAll(Console.WriteLine);
    }

    static int SleepAndReturn(int interval)
    {
      Thread.Sleep(interval*interval);
      return interval;
    }
  }
}

File: code/sorting/sleep_sort/sleep_sort.java
/* Part of Cosmos by OpenGenus Foundation */
import java.util.concurrent.CountDownLatch;
public class SleepSort {
	public static void sleepSortAndPrint(int[] nums) {
		final CountDownLatch doneSignal = new CountDownLatch(nums.length);
		for (final int num : nums) {
			new Thread(new Runnable() {
				public void run() {
					doneSignal.countDown();
					try {
						doneSignal.await();
 
						//using straight milliseconds produces unpredictable
						//results with small numbers
						//using 1000 here gives a nifty demonstration
						Thread.sleep(num * 1000);
						System.out.println(num);
					} catch (InterruptedException e) {
						e.printStackTrace();
					}
				}
			}).start();
		}
	}
	public static void main(String[] args) {
		int[] nums = new int[args.length];
		for (int i = 0; i < args.length; i++)
			nums[i] = Integer.parseInt(args[i]);
		sleepSortAndPrint(nums);
	}
}
